File: app/src/main/java/com/qurantv/app/domain/PageMapping.kt
package com.qurantv.app.domain

/**
 * Coordinate mapping between the SVG mushaf page space and screen pixels.
 *
 * The prompt verified `viewBox="0 0 235 235"` for early pages; live checking
 * shows other pages use different view boxes (e.g. page 187 is
 * `viewBox="0 0 345 550"`), so the actual viewBox is always parsed from the
 * SVG and used for scaling.
 */
data class ViewBox(val x: Float, val y: Float, val w: Float, val h: Float) {
    companion object {
        val DEFAULT = ViewBox(0f, 0f, 235f, 235f)
    }
}

object PageMapping {

    private val VIEW_BOX_REGEX = Regex("viewBox\\s*=\\s*[\"']([^\"']+)[\"']", RegexOption.IGNORE_CASE)

    fun parseViewBox(svg: String): ViewBox? {
        val match = VIEW_BOX_REGEX.find(svg) ?: return null
        val parts = match.groupValues[1]
            .trim()
            .split(Regex("[\\s,]+"))
            .mapNotNull { it.toFloatOrNull() }
        if (parts.size != 4) return null
        return ViewBox(parts[0], parts[1], parts[2], parts[3])
    }

    /** Maps a polygon from page space into a screen rect of (screenW x screenH). */
    fun toScreen(polygon: List<PointF>, viewBox: ViewBox, screenW: Float, screenH: Float): List<PointF> {
        if (screenW <= 0f || screenH <= 0f || viewBox.w <= 0f || viewBox.h <= 0f) return emptyList()
        val sx = screenW / viewBox.w
        val sy = screenH / viewBox.h
        return polygon.map { PointF((it.x - viewBox.x) * sx, (it.y - viewBox.y) * sy) }
    }
}

File: web/src/domain/PageMapping.ts
// Ported 1:1 from app/src/main/java/com/qurantv/app/domain/PageMapping.kt
// Coordinate mapping between the SVG mushaf page space and screen pixels.
// The actual viewBox is ALWAYS parsed from the SVG (pages vary: 235×235 early
// pages, page 187 is 345×550) — never assume 235.

export interface ViewBox {
  x: number;
  y: number;
  w: number;
  h: number;
}

export const DEFAULT_VIEW_BOX: ViewBox = { x: 0, y: 0, w: 235, h: 235 };

const VIEW_BOX_REGEX = /viewBox\s*=\s*["']([^"']+)["']/i;

export function parseViewBox(svg: string): ViewBox | null {
  const match = VIEW_BOX_REGEX.exec(svg);
  if (!match) return null;
  const parts = match[1]
    .trim()
    .split(/[\s,]+/)
    .map((it) => Number.parseFloat(it))
    .filter((n) => Number.isFinite(n));
  if (parts.length !== 4) return null;
  return { x: parts[0], y: parts[1], w: parts[2], h: parts[3] };
}

/** Maps a polygon from page space into a screen rect of (screenW x screenH). */
export function toScreen(
  polygon: PointF[],
  viewBox: ViewBox,
  screenW: number,
  screenH: number,
): PointF[] {
  if (screenW <= 0 || screenH <= 0 || viewBox.w <= 0 || viewBox.h <= 0) return [];
  const sx = screenW / viewBox.w;
  const sy = screenH / viewBox.h;
  return polygon.map((p) => ({ x: (p.x - viewBox.x) * sx, y: (p.y - viewBox.y) * sy }));
}

import type { PointF } from "./Models";
